File: nanopay/src/net/nanopay/meter/AdditionalDocumentsUpdatedIpHistoryDAO.java
package net.nanopay.meter;

import foam.core.FObject;
import foam.core.PropertyInfo;
import foam.core.X;
import foam.dao.DAO;
import foam.dao.ProxyDAO;
import foam.nanos.auth.User;
import foam.nanos.fs.File;

import javax.servlet.http.HttpServletRequest;
import java.util.Arrays;

public class AdditionalDocumentsUpdatedIpHistoryDAO extends ProxyDAO {
  public AdditionalDocumentsUpdatedIpHistoryDAO(X x, DAO delegate) {
    super(x, delegate);
  }

  @Override
  public FObject put_(X x, FObject obj) {
    User newUser = (User) obj;
    User oldUser = (User) getDelegate().find(newUser.getId());

    PropertyInfo prop = (PropertyInfo) User.getOwnClassInfo().getAxiomByName("additionalDocuments");
    Object[] newFiles = (Object[]) prop.get(newUser);
    if ( (oldUser == null && newFiles.length > 0)
      || (oldUser != null && !Arrays.deepEquals((Object[]) prop.get(oldUser), newFiles))
    ) {
      HttpServletRequest request = x.get(HttpServletRequest.class);
      String ipAddress = request.getRemoteAddr();

      IpHistory record = new IpHistory.Builder(x)
        .setUser(newUser.getId())
        .setIpAddress(ipAddress)
        .setDescription("Update additional documents").build();
      ((DAO) x.get("ipHistoryDAO")).put(record);
    }

    return super.put_(x, obj);
  }
}

File: nanopay/src/net/nanopay/meter/AdditionalDocumentsUpdatedIpHistoryDAO.js
/**
 * NANOPAY CONFIDENTIAL
 *
 * [2020] nanopay Corporation
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of nanopay Corporation.
 * The intellectual and technical concepts contained
 * herein are proprietary to nanopay Corporation
 * and may be covered by Canadian and Foreign Patents, patents
 * in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from nanopay Corporation.
 */

foam.CLASS({
  package: 'net.nanopay.meter',
  name: 'AdditionalDocumentsUpdatedIpHistoryDAO',
  extends: 'foam.dao.ProxyDAO',

  documentation: `Decorating DAO for capturing user IP address when
      additional documents is updated.`,

  javaImports: [
    'foam.nanos.auth.User',
    'foam.nanos.fs.File',
    'java.util.Arrays'
  ],

  methods: [
    {
      name: 'put_',
      javaCode: `
        User newUser = (User) obj;
        User oldUser = (User) getDelegate().find(newUser.getId());

        File[] newFiles = newUser.getAdditionalDocuments();
        File[] oldFiles = null;
        if (oldUser != null) {
          oldFiles = oldUser.getAdditionalDocuments();
        }

        if (oldUser != null && !Arrays.deepEquals(oldFiles, newFiles)) {
          IpHistoryService ipHistoryService = new IpHistoryService(x);
          String description = String.format("Upload:%s additional documents",
            getUploadAction(oldFiles.length, newFiles.length));

          ipHistoryService.record(newUser, description);
        }

        return super.put_(x, obj);
      `
    },
    {
      name: 'getUploadAction',
      type: 'String',
      args: [
        { type: 'Integer', name: 'o' },
        { type: 'Integer', name: 'n' }
      ],
      javaCode: `
        if (n > o) return "add";
        else if (o > n) return "delete";
        return "update";
      `
    }
  ]
});
